### PR TITLE
executor: Fix the display of the prepared sql 

### LINKED ISSUE #16062: The slow log of `Prepare` statements don't record the real SQL
## Description
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. What did you do?
The slow log of `Prepare` statements don't record the real SQL . And it's not similar with MySQL .

```sql

MySQL [test]> PREPARE mystmt FROM 'SELECT sleep(?)';
Query OK, 0 rows affected (0.01 sec)

MySQL [test]> SET @number = 5;
Query OK, 0 rows affected (0.00 sec)

MySQL [test]> EXECUTE mystmt USING @number;
+----------+
| sleep(?) |
+----------+
|        0 |
+----------+
1 row in set (5.00 sec)

MySQL [test]> select tidb_version();
+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| tidb_version()                                                                                                                                                                                                                                                                                                 |
+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| Release Version: v2.1.17
Git Commit Hash: 603f0b95d391c336132f33de6877feb6e6984e54
Git Branch: HEAD
UTC Build Time: 2020-04-02 10:31:39
GoVersion: go version go1.13.6 linux/amd64
Race Enabled: false
TiKV Min Version: 2.1.0-alpha.1-ff3dd160846b7d1aed9079c389fc188f7f5ea13e
Check Table Before Drop: false |
+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
1 row in set (0.00 sec)
```

<!-- If possible, provide a recipe for reproducing the error. -->

### 2. What did you expect to see?
**In Mysql:**
```sql
mysql> PREPARE mystmt FROM 'SELECT sleep(?)';
Query OK, 0 rows affected (0.00 sec)
Statement prepared

mysql> SET @number = 5;
Query OK, 0 rows affected (0.00 sec)

mysql> EXECUTE mystmt USING @number;
+----------+
| sleep(?) |
+----------+
|        0 |
+----------+
1 row in set (5.00 sec)

mysql> \! tail -n 10 /home/mysql/mysqllog/mysql-slow.log
# Time: 2020-04-02T22:32:17.643358-05:00
# User@Host: root[root] @  [127.0.0.1]  Id:  1572
# Query_time: 5.000391  Lock_time: 0.000000 Rows_sent: 1  Rows_examined: 0
SET timestamp=1585881137;
SELECT sleep(5);
# Time: 2020-04-03T06:15:05.959133-05:00
# User@Host: root[root] @  [172.16.4.142]  Id:  1573
# Query_time: 5.000401  Lock_time: 0.000000 Rows_sent: 1  Rows_examined: 0
SET timestamp=1585908905;
SELECT sleep(5);
mysql> select @@version;
+------------+
| @@version  |
+------------+
| 5.7.23-log |
+------------+
1 row in set (0.00 sec)

```
### 3. What did you see instead?
**In TiDB :**
<img width="927" alt="企业微信截图_6fb4caaf-5c4e-4dcd-a162-9e06da06417e" src="https://user-images.githubusercontent.com/20888467/78349160-bdbd3900-75d5-11ea-9c3d-cba46db02307.png">

### 4. What version of TiDB are you using? (`tidb-server -V` or run `select tidb_version();` on TiDB)
v2.1.17



## SIG slack channel
[#sig-exec](https://tidbcommunity.slack.com/archives/CMRD79DRR)

## Score
- 300

## Mentor
 * @qw4990

### LINKED ISSUE #16653: statements_summary query_sample_text shows prepared statement form
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. What did you do?

I ran tpcc against master, and then observed the workload in `information_schema.statements_summary`.

### 2. What did you expect to see?

I was really hoping to see `query_sample_text` so that I could take the query and run it through explain. Instead it showed only the prepared statement form - which is non-explainable.

### 3. What did you see instead?

```
mysql> select * from statements_summary where stmt_type='select' order by sum_latency desc limit 2\G
*************************** 1. row ***************************
       SUMMARY_BEGIN_TIME: 2020-04-20 20:30:00
         SUMMARY_END_TIME: 2020-04-20 21:00:00
                STMT_TYPE: select
              SCHEMA_NAME: tpcc
                   DIGEST: 79cdf6c3eccead898330bda0b5630d505010bed8dac509c9d1a4e485e9e29b60
              DIGEST_TEXT: select c_discount , c_last , c_credit , w_tax from customer , warehouse where w_id = ? and c_w_id = w_id and c_d_id = ? and c_id = ?
              TABLE_NAMES: tpcc.customer,tpcc.warehouse
              INDEX_NAMES: NULL
              SAMPLE_USER: root
               EXEC_COUNT: 16146
               SUM_ERRORS: 0
             SUM_WARNINGS: 0
              SUM_LATENCY: 36927867322
              MAX_LATENCY: 104421632
              MIN_LATENCY: 471610
              AVG_LATENCY: 2287121
        AVG_PARSE_LATENCY: 9622
        MAX_PARSE_LATENCY: 4654548
      AVG_COMPILE_LATENCY: 318677
      MAX_COMPILE_LATENCY: 5278069
         SUM_COP_TASK_NUM: 0
     MAX_COP_PROCESS_TIME: 0
  MAX_COP_PROCESS_ADDRESS: NULL
        MAX_COP_WAIT_TIME: 0
     MAX_COP_WAIT_ADDRESS: NULL
         AVG_PROCESS_TIME: 0
         MAX_PROCESS_TIME: 0
            AVG_WAIT_TIME: 0
            MAX_WAIT_TIME: 0
         AVG_BACKOFF_TIME: 0
         MAX_BACKOFF_TIME: 0
           AVG_TOTAL_KEYS: 0
           MAX_TOTAL_KEYS: 0
       AVG_PROCESSED_KEYS: 0
       MAX_PROCESSED_KEYS: 0
        AVG_PREWRITE_TIME: 0
        MAX_PREWRITE_TIME: 0
          AVG_COMMIT_TIME: 0
          MAX_COMMIT_TIME: 0
   AVG_GET_COMMIT_TS_TIME: 0
   MAX_GET_COMMIT_TS_TIME: 0
  AVG_COMMIT_BACKOFF_TIME: 0
  MAX_COMMIT_BACKOFF_TIME: 0
    AVG_RESOLVE_LOCK_TIME: 0
    MAX_RESOLVE_LOCK_TIME: 0
AVG_LOCAL_LATCH_WAIT_TIME: 0
MAX_LOCAL_LATCH_WAIT_TIME: 0
           AVG_WRITE_KEYS: 0
           MAX_WRITE_KEYS: 0
           AVG_WRITE_SIZE: 0
           MAX_WRITE_SIZE: 0
     AVG_PREWRITE_REGIONS: 0
     MAX_PREWRITE_REGIONS: 0
            AVG_TXN_RETRY: 0
            MAX_TXN_RETRY: 0
        SUM_BACKOFF_TIMES: 0
            BACKOFF_TYPES: NULL
                  AVG_MEM: 53512
                  MAX_MEM: 53512
        AVG_AFFECTED_ROWS: 0
               FIRST_SEEN: 2020-04-20 20:49:34
                LAST_SEEN: 2020-04-20 20:51:25
        QUERY_SAMPLE_TEXT: SELECT c_discount, c_last, c_credit, w_tax FROM customer, warehouse WHERE w_id = ? AND c_w_id = w_id AND c_d_id = ? AND c_id = ?
         PREV_SAMPLE_TEXT: 
              PLAN_DIGEST: d40707dd740e8032fe6a362a242c86f5bb6409008ae3e968c45833be2f8e3065
                     PLAN: 	Projection_7    	root	0.9999999999999999	tpcc.customer.c_discount, tpcc.customer.c_last, tpcc.customer.c_credit, tpcc.warehouse.w_tax
	└─HashJoin_8    	root	0.9999999999999999	CARTESIAN inner join
	  ├─Point_Get_10	root	1                 	table:customer, index:PRIMARY(c_w_id, c_d_id, c_id)
	  └─Point_Get_11	root	1                 	table:warehouse, handle:10

```

See the `?` for values under `query_sample_text`. In MySQL the query sample shows something that is directly explainable. The digest_text shows the prepared statement form.

### 4. What version of TiDB are you using? (`tidb-server -V` or run `select tidb_version();` on TiDB)

```
mysql> select tidb_version()\G
*************************** 1. row ***************************
tidb_version(): Release Version: v4.0.0-beta.2-302-g8c455e86c
Git Commit Hash: 8c455e86c136ac020129998c94f890448367f62a
Git Branch: master
UTC Build Time: 2020-04-21 02:25:12
GoVersion: go1.13.8
Race Enabled: false
TiKV Min Version: v3.0.0-60965b006877ca7234adaced7890d7b029ed1306
Check Table Before Drop: false
1 row in set (0.00 sec)
```

### PR DESCRIPTION
**Pull request by mmyj**
### What problem does this PR solve?

Issue Number: fixes #16062 fixes #16653

Problem Summary:
`Prepare` statements don't record/display the real SQL in the follow table/command
1. SLOW_QUERY
2. STATEMENTS_SUMMARY
3. SHOW PROCESSLIST

### What is changed and how it works?

What's Changed:

Modify the `FormatSQL` function, replace `?` to the prepared params.

Tests 

- Unit test

### Release note 

* The slow log, `show processlist` and `statements_summary` table now shows the actual statement for prepared statements, rather than the prepared form.

### CODE REVIEW

**Comment by wwar** [comment]
I believe this may also fix https://github.com/pingcap/tidb/issues/16653 ?

Hoping so! It is quite annoying to not be able to explain the queries from statements_summary.

**Comment by wwar** [comment]
**Edit:** It doesn't look like it fixes #16653 

I found a bug while testing this. To reproduce, run `tiup bench tpcc prepare` against the tidb-server (does not require pd or tikv):

```
[2020/06/26 20:24:51.299 -06:00] [ERROR] [conn.go:668] ["connection running loop panic"] [conn=2] [lastSQL="SELECT count(*) \n\tFROM (  SELECT  c.c_id, c.c_d_id, c.c_w_id, c.c_balance c1, \n\t\t\t\t   (SELECT sum(ol_amount) FROM orders STRAIGHT_JOIN order_line \n\t\t\t\t\t WHERE OL_W_ID=O_W_ID \n\t\t\t\t\t   AND OL_D_ID = O_D_ID \n\t\t\t\t\t   AND OL_O_ID = O_ID \n\t\t\t\t\t   AND OL_DELIVERY_D IS NOT NULL \n\t\t\t\t\t   AND O_W_ID=? \n\t\t\t\t\t   AND O_D_ID=c.C_D_ID \n\t\t\t\t\t   AND O_C_ID=c.C_ID) sm, (SELECT  sum(h_amount)  from  history \n\t\t\t\t\t\t\t\t\t\t\t\tWHERE H_C_W_ID=? \n\t\t\t\t\t\t\t\t\t\t\t\t  AND H_C_D_ID=c.C_D_ID \n\t\t\t\t\t\t\t\t\t\t\t\t  AND H_C_ID=c.C_ID) smh \n\t\t\t FROM customer c \n\t\t\tWHERE  c.c_w_id = ? ) t\n   WHERE c1<>sm-smh [arguments: (1, 1, 1)]"] [err="runtime error: index out of range [3] with length 3"] [stack="goroutine 1515 [running]:\ngithub.com/pingcap/tidb/server.(*clientConn).Run.func1(0x3acbc20, 0xc0129c5590, 0xc0129c2480)\n\t/tmp/tidb/server/conn.go:666 +0xee\npanic(0x33335c0, 0xc013814b00)\n\t/usr/lib/go-1.13/src/runtime/panic.go:679 +0x1b2\ngithub.com/pingcap/tidb/executor.FormatSQL.func1(0x3b11b60, 0xc0128e8800)\n\t/tmp/tidb/executor/adapter.go:781 +0x464\ngithub.com/pingcap/tidb/util/stringutil.StringerFunc.String(...)\n\t/tmp/tidb/util/stringutil/string_util.go:280\ngithub.com/pingcap/tidb/executor.(*ExecStmt).LogSlowQuery(0xc011b03440, 0x5cbce5df39c0000, 0xc00b9f0001)\n\t/tmp/tidb/executor/adapter.go:870 +0x3e9\ngithub.com/pingcap/tidb/executor.(*ExecStmt).FinishExecuteStmt(0xc011b03440, 0x5cbce5df39c0000, 0x13e0001)\n\t/tmp/tidb/executor/adapter.go:800 +0x4b\ngithub.com/pingcap/tidb/executor.(*ExecStmt).CloseRecordSet(0xc011b03440, 0x5cbce5df39c0000, 0x0, 0x0)\n\t/tmp/tidb/executor/adapter.go:815 +0x4b\ngithub.com/pingcap/tidb/executor.(*recordSet).Close(0xc00b995130, 0xc010193e58, 0xc012a1c000)\n\t/tmp/tidb/executor/adapter.go:153 +0x79\ngithub.com/pingcap/tidb/session.(*execStmtResult).Close(0xc012db30e0, 0x4000, 0x4f)\n\t/tmp/tidb/session/session.go:1246 +0x40\ngithub.com/pingcap/tidb/server.(*tidbResultSet).Close(0xc00b995180, 0x400, 0x0)\n\t/tmp/tidb/server/driver_tidb.go:302 +0x4a\ngithub.com/pingcap/parser/terror.Call(0xc00f52b3a8)\n\t/home/wwar/go/pkg/mod/github.com/pingcap/parser@v0.0.0-20200623082809-b74301ac298b/terror/terror.go:354 +0x3c\ngithub.com/pingcap/tidb/server.(*clientConn).handleStmtExecute(0xc0129c2480, 0x3acbc20, 0xc01426efc0, 0xc00454f381, 0x29, 0x29, 0x0, 0x0)\n\t/tmp/tidb/server/conn_stmt.go:215 +0x796\ngithub.com/pingcap/tidb/server.(*clientConn).dispatch(0xc0129c2480, 0x3acbc20, 0xc0129c5590, 0xc00454f381, 0x2a, 0x29, 0x0, 0x0)\n\t/tmp/tidb/server/conn.go:920 +0x80f\ngithub.com/pingcap/tidb/server.(*clientConn).Run(0xc0129c2480, 0x3acbc20, 0xc0129c5590)\n\t/tmp/tidb/server/conn.go:721 +0x27f\ngithub.com/pingcap/tidb/server.(*Server).onConn(0xc010180370, 0xc0129c2480)\n\t/tmp/tidb/server/server.go:415 +0xb18\ncreated by github.com/pingcap/tidb/server.(*Server).Run\n\t/tmp/tidb/server/server.go:333 +0x709\n"]

```

**Comment by mmyj** [comment]
> I believe this may also fix #16653 ?
> 
> Hoping so! It is quite annoying to not be able to explain the queries from statements_summary.

This bugfix can not solve #16653, but I will fix it.

**Comment by wwar** [comment]
> This bugfix can not solve #16653, but I will fix it.

Thanks! Not to scope creep on you, but I also noticed that in MySQL the `Info` column of `SHOW PROCESSLIST` will also show the query with values in it.

**Comment by mmyj** [comment]
> > This bugfix can not solve #16653, but I will fix it.
> 
> Thanks! Not to scope creep on you, but I also noticed that in MySQL the `Info` column of `SHOW PROCESSLIST` will also show the query with values in it.

I had reproduced it, using follow sql.
```
prepare mysleep from 'select sleep(?)';
set @a = 10;
execute mysleep using @a;
```

In TiDB, the info column  is `execute mysleep using @a`
In MySQL, the info column  is `select sleep(10)`

**Comment by mmyj** [comment]
> > This bugfix can not solve #16653, but I will fix it.
> 
> Thanks! Not to scope creep on you, but I also noticed that in MySQL the `Info` column of `SHOW PROCESSLIST` will also show the query with values in it.

Now the `Info` column of `SHOW PROCESSLIST` will show the query with values in it.

File: executor/adapter.go [hunk 1]
```diff
@@ -764,8 +767,28 @@ func FormatSQL(sql string, pps variable.PreparedParams) stringutil.StringerFunc
 		if maxQueryLen := atomic.LoadUint64(&cfg.Log.QueryLogMaxLen); uint64(length) > maxQueryLen {
 			sql = fmt.Sprintf("%.*q(len:%d)", maxQueryLen, sql, length)
 		}
-		return QueryReplacer.Replace(sql) + pps.String()
+		replaceSQL := QueryReplacer.Replace(sql)
+		if len(pps) == 0 {
+			return replaceSQL
+		}
+		return formatPreparedSQL(replaceSQL, pps)
```
Comment (qw4990) [inline executor/adapter.go:774]: Should we format prepared SQLs before truncating them? How about moving this `formatPreparedSQL` to line 765?

Comment (mmyj) [reply executor/adapter.go:774]: I will fix it.

File: executor/adapter.go [hunk 2]
```diff
@@ -759,13 +762,32 @@ func (a *ExecStmt) logAudit() {
...
+func formatPreparedSQL(sql string, pps variable.PreparedParams) string {
+	sqlBuffer := bytes.NewBuffer([]byte{})
+	paramsIndex := 0
+	for i := 0; i < len(sql); i++ {
+		c := sql[i]
+		if c != '?' {
+			sqlBuffer.WriteByte(c)
+			continue
+		}
+		sqlBuffer.WriteString(fmt.Sprintf("%v", pps[paramsIndex].GetValue()))
```
Comment (SunRunAway) [inline executor/adapter.go:787]: I think we should handle the format of null value and string value, for example, add quotes around a string(`abc`->`"abc"`).
You may take a reference of  `types.DatumsToString` .

File: executor/adapter.go [hunk 3]
```diff
@@ -988,6 +1010,8 @@ func (a *ExecStmt) SummaryStmt(succ bool) {
 	sql := a.Text
 	if sensitiveStmt, ok := a.StmtNode.(ast.SensitiveStmtNode); ok {
 		sql = sensitiveStmt.SecureText()
+	} else if len(sessVars.PreparedParams) > 0 {
+		sql = formatPreparedSQL(sessVars.StmtCtx.OriginalSQL, sessVars.PreparedParams)
```
Comment (SunRunAway) [inline executor/adapter.go:1014]: Could FormatSQL also be used here?

File: executor/adapter_test.go [hunk 4]
```diff
@@ -35,3 +41,40 @@ func (s *testSuiteP2) TestQueryTime(c *C) {
...
+		types.NewIntDatum(1),
+		types.NewFloat64Datum(2),
+		types.NewStringDatum("\"hello, 世界\""),
+		types.NewStringDatum("[1, 2, 3]"),
+		types.NewStringDatum("{}"),
+		types.NewStringDatum(`{"a": "9223372036854775809"}`),
+		mustParseTimeIntoDatum("2011-11-10 11:11:11.111111", mysql.TypeTimestamp, 6),
+	}
+	preparedSQL := executor.FormatSQL("select ?, ?, ?, ?, ?, ?, ?;", preparedParams)()
+	c.Check(preparedSQL, Equals, "select 1, 2, \"hello, 世界\", [1, 2, 3], {}, {\"a\": \"9223372036854775809\"}, 2011-11-10 11:11:11.111111;")
```
Comment (SunRunAway) [inline executor/adapter_test.go:56]: I think string in select field should be quoted.
Right: `select "abc";`
Wrong: `select abc;`

**Comment by SunRunAway** [comment]
> goroutine 1515 [running]:\ngithub.com/pingcap/tidb/server.(*clientConn).Run.func1(0x3acbc20, 0xc0129c5590, 0xc0129c2480)\n\t/tmp/tidb/server/conn.go:666 +0xee\npanic(0x33335c0, 0xc013814b00)\n\t/usr/lib/go-1.13/src/runtime/panic.go:679 +0x1b2\ngithub.com/pingcap/tidb/executor.FormatSQL.func1(0x3b11b60, 0xc0128e8800)\n\t/tmp/tidb/executor/adapter.go:781 +0x464\ngithub.com/pingcap/tidb/util/stringutil.StringerFunc.String(...)\n\t/tmp/tidb/util/stringutil/string_util.go:280\ngithub.com/pingcap/tidb/executor.(*ExecStmt).LogSlowQuery(0xc011b03440, 0x5cbce5df39c0000, 0xc00b9f0001)\n\t/tmp/tidb/executor/adapter.go:870 +0x3e9\ngithub.com/pingcap/tidb/executor.(*ExecStmt).FinishExecuteStmt(0xc011b03440, 0x5cbce5df39c0000, 0x13e0001)\n\t/tmp/tidb/executor/adapter.go:800 +0x4b\ngithub.com/pingcap/tidb/executor.(*ExecStmt).CloseRecordSet(0xc011b03440, 0x5cbce5df39c0000, 0x0, 0x0)\n\t/tmp/tidb/executor/adapter.go:815 +0x4b\ngithub.com/pingcap/tidb/executor.(*recordSet).Close(0xc00b995130, 0xc010193e58, 0xc012a1c000)\n\t/tmp/tidb/executor/adapter.go:153 +0x79\ngithub.com/pingcap/tidb/session.(*execStmtResult).Close(0xc012db30e0, 0x4000, 0x4f)\n\t/tmp/tidb/session/session.go:1246 +0x40\ngithub.com/pingcap/tidb/server.(*tidbResultSet).Close(0xc00b995180, 0x400, 0x0)\n\t/tmp/tidb/server/driver_tidb.go:302 +0x4a\ngithub.com/pingcap/parser/terror.Call(0xc00f52b3a8)\n\t/home/wwar/go/pkg/mod/github.com/pingcap/parser@v0.0.0-20200623082809-b74301ac298b/terror/terror.go:354 +0x3c\ngithub.com/pingcap/tidb/server.(*clientConn).handleStmtExecute(0xc0129c2480, 0x3acbc20, 0xc01426efc0, 0xc00454f381, 0x29, 0x29, 0x0, 0x0)\n\t/tmp/tidb/server/conn_stmt.go:215 +0x796\ngithub.com/pingcap/tidb/server.(*clientConn).dispatch(0xc0129c2480, 0x3acbc20, 0xc0129c5590, 0xc00454f381, 0x2a, 0x29, 0x0, 0x0)\n\t/tmp/tidb/server/conn.go:920 +0x80f\ngithub.com/pingcap/tidb/server.(*clientConn).Run(0xc0129c2480, 0x3acbc20, 0xc0129c5590)\n\t/tmp/tidb/server/conn.go:721 +0x27f\ngithub.com/pingcap/tidb/server.(*Server).onConn(0xc010180370, 0xc0129c2480)\n\t/tmp/tidb/server/server.go:415 +0xb18\ncreated by github.com/pingcap/tidb/server.(*Server).Run\n\t/tmp/tidb/server/server.go:333 +0x709\n"

And is this panic be resolved? @mmyj

**Comment by wwar** [comment]
Yes, I can confirm the panic is resolved.

**Comment by mmyj** [comment]
> > goroutine 1515 [running]:\ngithub.com/pingcap/tidb/server.(*clientConn).Run.func1(0x3acbc20, 0xc0129c5590, 0xc0129c2480)\n\t/tmp/tidb/server/conn.go:666 +0xee\npanic(0x33335c0, 0xc013814b00)\n\t/usr/lib/go-1.13/src/runtime/panic.go:679 +0x1b2\ngithub.com/pingcap/tidb/executor.FormatSQL.func1(0x3b11b60, 0xc0128e8800)\n\t/tmp/tidb/executor/adapter.go:781 +0x464\ngithub.com/pingcap/tidb/util/stringutil.StringerFunc.String(...)\n\t/tmp/tidb/util/stringutil/string_util.go:280\ngithub.com/pingcap/tidb/executor.(*ExecStmt).LogSlowQuery(0xc011b03440, 0x5cbce5df39c0000, 0xc00b9f0001)\n\t/tmp/tidb/executor/adapter.go:870 +0x3e9\ngithub.com/pingcap/tidb/executor.(*ExecStmt).FinishExecuteStmt(0xc011b03440, 0x5cbce5df39c0000, 0x13e0001)\n\t/tmp/tidb/executor/adapter.go:800 +0x4b\ngithub.com/pingcap/tidb/executor.(*ExecStmt).CloseRecordSet(0xc011b03440, 0x5cbce5df39c0000, 0x0, 0x0)\n\t/tmp/tidb/executor/adapter.go:815 +0x4b\ngithub.com/pingcap/tidb/executor.(*recordSet).Close(0xc00b995130, 0xc010193e58, 0xc012a1c000)\n\t/tmp/tidb/executor/adapter.go:153 +0x79\ngithub.com/pingcap/tidb/session.(*execStmtResult).Close(0xc012db30e0, 0x4000, 0x4f)\n\t/tmp/tidb/session/session.go:1246 +0x40\ngithub.com/pingcap/tidb/server.(*tidbResultSet).Close(0xc00b995180, 0x400, 0x0)\n\t/tmp/tidb/server/driver_tidb.go:302 +0x4a\ngithub.com/pingcap/parser/terror.Call(0xc00f52b3a8)\n\t/home/wwar/go/pkg/mod/github.com/pingcap/parser@v0.0.0-20200623082809-b74301ac298b/terror/terror.go:354 +0x3c\ngithub.com/pingcap/tidb/server.(*clientConn).handleStmtExecute(0xc0129c2480, 0x3acbc20, 0xc01426efc0, 0xc00454f381, 0x29, 0x29, 0x0, 0x0)\n\t/tmp/tidb/server/conn_stmt.go:215 +0x796\ngithub.com/pingcap/tidb/server.(*clientConn).dispatch(0xc0129c2480, 0x3acbc20, 0xc0129c5590, 0xc00454f381, 0x2a, 0x29, 0x0, 0x0)\n\t/tmp/tidb/server/conn.go:920 +0x80f\ngithub.com/pingcap/tidb/server.(*clientConn).Run(0xc0129c2480, 0x3acbc20, 0xc0129c5590)\n\t/tmp/tidb/server/conn.go:721 +0x27f\ngithub.com/pingcap/tidb/server.(*Server).onConn(0xc010180370, 0xc0129c2480)\n\t/tmp/tidb/server/server.go:415 +0xb18\ncreated by github.com/pingcap/tidb/server.(*Server).Run\n\t/tmp/tidb/server/server.go:333 +0x709\n"
> 
> And is this panic be resolved? @mmyj

This panic is resolved.

**Comment by nullnotnil** [comment]
@mmyj Please resolve conflicts.

**Comment by SunRunAway** [comment]
> @mmyj Please resolve conflicts.

And address this review https://github.com/pingcap/tidb/pull/18218#pullrequestreview-445574565

File: executor/adapter.go (hunk 2 above)
Comment (mmyj) [reply executor/adapter.go:787]: i will fix it

File: executor/adapter_test.go (hunk 4 above)
Comment (mmyj) [reply executor/adapter_test.go]: I will fix it.

**Comment by mmyj** [comment]
PTAL @SunRunAway

File: executor/adapter.go [hunk 5]
```diff
@@ -773,13 +776,37 @@ func (a *ExecStmt) logAudit() {
...
+	}
+}
+
+// formatPreparedSQL is used to replace the prepared params for original SQL
+func formatPreparedSQL(sql string, pps variable.PreparedParams) string {
+	sqlBuffer := bytes.NewBuffer([]byte{})
+	paramsIndex := 0
+	for i := 0; i < len(sql); i++ {
+		c := sql[i]
+		if c != '?' {
```
[the remark this answers is not in the document]
Comment (mmyj) [reply executor/adapter.go:797]: It will go wrong. T_T

**Comment by zz-jason** [comment]
I'm going to close this PR since it hasn't been updated for a long time. feel free to reopen if you want to continue with it. thank you for your contribution.

**Pull request closed without merging**